### PR TITLE
tests: add smoke test for nydusd hot upgrade

### PR DESCRIPTION
**Pull request by loheagn**
## Relevant Issue (if applicable)

https://github.com/dragonflyoss/nydus/issues/1421

## Details

1. add smoke test for nydusd hot upgrade
2. rename the `failover-test` to `takeover-test` as the `takeover-test` contains tests for both failover and hot upgrade

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: smoke/tests/tool/snapshotter.go [hunk 1]
```diff
@@ -64,6 +73,34 @@ func NewSnapshotterClient(sock string) *SnapshotterClient {
...
+		reqJSON, err = json.Marshal(body)
+		if err != nil {
+			return nil, errors.Wrap(err, "marshal request body")
+		}
+	} else {
+		reqJSON = make([]byte, 0)
+	}
+
+	url := fmt.Sprintf("http://unix%s", urlSuffix)
+	req, err := http.NewRequest(method, url, bytes.NewBuffer(reqJSON))
```
Comment (imeoer) [inline smoke/tests/tool/snapshotter.go:88]: Maybe give a `nil` if paasing an empty body.

File: smoke/tests/tool/snapshotter.go [hunk 2]
```diff
@@ -64,6 +73,34 @@ func NewSnapshotterClient(sock string) *SnapshotterClient {
...
+	if err != nil {
+		return nil, errors.Wrap(err, "build request")
+	}
+	resp, err := cli.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "do request")
+	}
```
Comment (imeoer) [inline smoke/tests/tool/snapshotter.go:95]: Should move `defer resp.Body.Close()` to here, and print the possible error message from `io.ReadAll(resp.Body)` if we encounter the abnormal status code except 2xx.

File: smoke/tests/takeover_test.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		killCmd := fmt.Sprintf("kill -9 %d", daemon.Pid)
+		tool.Run(t, killCmd)
+	}
+
+	// wait for the nydus daemons recover
+	time.Sleep(5 * time.Second)
+
+	// check the container by requesting its wait url
+	runArgs := tool.GetRunArgs(t, imageName)
+	resp, err := http.Get(runArgs.WaitURL)
```
Comment (imeoer) [inline smoke/tests/takeover_test.go:85]: Should do `defer resp.Close()` if no `err`. :)

File: smoke/tests/takeover_test.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	if err := f.snapshotterCli.Upgrade(upgradeReq); err != nil {
+		t.Fatalf("failed to upgrade: %s", err.Error())
+	}
+
+	// wait for the nydus daemons recover
+	time.Sleep(5 * time.Second)
+
+	// check the container by requesting its wait url
+	runArgs := tool.GetRunArgs(t, imageName)
+	resp, err := http.Get(runArgs.WaitURL)
```
Comment (imeoer) [inline smoke/tests/takeover_test.go:124]: Ditto.

File: smoke/tests/takeover_test.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+func (f *TakeoverTestSuit) TestFailover(t *testing.T) {
+	imageName := f.testImage
+
+	containerName := uuid.NewString()
+	tool.RunContainerSimple(t, imageName, snapshotter, containerName, false)
+	defer f.rmContainer(containerName)
+
+	daemons, err := f.snapshotterCli.GetNydusDaemonInfos()
+	if err != nil {
```
Comment (imeoer) [inline smoke/tests/takeover_test.go:70]: Let's use `require.NoError(t, err, "xxx")` to simplify all error checks.

**Review by imeoer — COMMENTED** [review]
Thanks for the work!

**Comment by imeoer** [comment]
@loheagn Can we also give a doc to explain how the failover & hot upgrade work and give the usage? Thanks!

**Comment by loheagn** [comment]
> @loheagn Can we also give a doc to explain how the failover & hot upgrade work and give the usage? For example `docs/nydus-failover-upgrade.md`, Thanks!

@imeoer Thanks for the review! All comments are fixed. I will submit another PR to add the doc.

**Comment by Desiki-high** [comment]
@loheagn Please user rebase to update branch not `Merge branch 'master'`.

**Comment by loheagn** [comment]
> @loheagn Please user rebase to update branch not `Merge branch 'master'`.

fixed

**Review by imeoer — APPROVED** [review]
Thanks!

**Pull request merged**